TheThingsNetwork/lorawan-stack
# Allow re-request of validation email before token expires

#### Summary



Closes #6525 

#### Changes



- Add `EmailInterval` options to user registration configuration.
- Update request validation method
- Add unit tests for contact_registry
- add test cases for contact_validation_store.
- Update `validations_already_sent` error, now includes `retry_interval`

#### Testing



Manually tested, unit tests.

<details>
  <summary> Test steps </summary>
   
   Run Stack locally with `contact-info-validation.required` as `true` and a `retry-interval` with a low value.

   Config example:
   ```yaml
   is:
     user-registration:
      contact-info-validation:
        required: false
        token-ttl: 48h0m0s
        retry-interval: 0h2m0s
  ```

Perform the following steps:

  1. Create a new non admin user. `ttn-lw-cli usr create <usr_id> --primary-email-address <email> --password <password>`
  2. Request validation `ttn-lw-cli usr contact-info request-validation <usr_id>`
    - Should return an error
  3. Wait until the time set on `retry-interval` has passed.
  4. Request another validation `ttn-lw-cli usr contact-info request-validation <usr_id>`
    

  The last validation request should've worked, as `token-ttl` is the default or a lot larger than the `retry-interval` it means that a new validation email was sent without reseting the validation itself.

  If you want to be even more detailed on the testing you could enable email being written to the local directory and then making sure that there is a new email after requesting a validation. The config example for doing such behavior is listed below, the steps are the same as the ones mentioned before you just need to count the amount of email before and after the validation request.

  ```yaml
  tls:
    source: file
    root-ca: ./ca.pem
    certificate: ./cert.pem
    key: ./key.pem

  is:
    email:
      dir: ".env/emails" # Example of directory path
      provider: "dir"
      sender-address: ""
      sender-name: # Example"employee@thethingsindustries.com"
      templates:
        directory: # Example "<repo_path>/pkg/email/templates"
    user-registration:
      contact-info-validation:
        required: true
        token-ttl: 48h0m0s
        retry-interval: 0h2m0s
  ```
</details>

##### Regressions



This adds the ability to resend the same validation, instead of waiting for the validation to expire in order to send a new one,  I took care to guarantee that the change doesn't break what was the expected behavior.

#### Notes for Reviewers



https://github.com/TheThingsNetwork/lorawan-stack/issues/6567 Is the main reason as of why I did the updates that are related to the contact_info store and registry, its counter intuitive since the field is deprecated but since this is specifically around the behavior of validations it felt appropriate to write these tests. They will change after separating the contact_info from validations. 

#### Checklist
